packages/material-ui-lab/src/TreeView/descendants.js

```diff
@@ -0,0 +1,214 @@
+    ...Object.values(descendant),
+  ]);
+
+  return { parentId, index };
+}
+
+export function useDescendantsInit() {
+  return React.useState([]);
+}
+
+export function DescendantProvider(props) {
+  const { children, items, set, id } = props;
+
+  const registerDescendant = React.useCallback(
+    ({ element, ...rest }) => {
+      if (!element) {
+        return;
+      }
+
+      set((oldItems) => {
+        let newItems;
+        if (oldItems.length === 0) {
+          // If there are no items, register at index 0 and bail.
+          newItems = [
+            ...oldItems,
+            {
+              ...rest,
+              element,
+              index: 0,
+            },
+          ];
+        } else if (oldItems.find((item) => item.element === element)) {
+          // If the element is already registered, just use the same array
+          newItems = oldItems;
+        } else {
+          // When registering a descendant, we need to make sure we insert in
+          // into the array in the same order that it appears in the DOM. So as
+          // new descendants are added or maybe some are removed, we always know
+          // that the array is up-to-date and correct.
+          //
+          // So here we look at our registered descendants and see if the new
+          // element we are adding appears earlier than an existing descendant's
+          // DOM node via `node.compareDocumentPosition`. If it does, we insert
+          // the new element at this index. Because `registerDescendant` will be
+          // called in an effect every time the descendants state value changes,
+          // we should be sure that this index is accurate when descendent
+          // elements come or go from our component.
+          const index = findIndex(oldItems, (item) => {
+            if (!item.element || !element) {
+              return false;
+            }
+            // Does this element's DOM node appear before another item in the
+            // array in our DOM tree? If so, return true to grab the index at
+            // this point in the array so we know where to insert the new
+            // element.
+            return Boolean(
+              // eslint-disable-next-line no-bitwise
+              item.element.compareDocumentPosition(element) & Node.DOCUMENT_POSITION_PRECEDING,
+            );
+          });
+
+          const newItem = {
+            ...rest,
+            element,
+            index,
+          };
+
+          // If an index is not found we will push the element to the end.
+          if (index === -1) {
+            newItems = [...oldItems, newItem];
+          } else {
+            newItems = [...oldItems.slice(0, index), newItem, ...oldItems.slice(index)];
+          }
+        }
+        return newItems.map((item, index) => ({ ...item, index }));
+      });
+    },
+    [set],
+  );
+
+  const unregisterDescendant = React.useCallback(
+    (element) => {
+      if (!element) {
+        return;
+      }
```

Comment:
This defensive logic seems weird. It seems that it should be handled before calling this method.

Reply:
Hmm, I know what you mean. I'll change it in the focus changes so it doesn't cause a merge conflict: